### PR TITLE
new check: MisplacedWeakBlocker

### PR DESCRIPTION
**Pull request by arthurzam**
Resolves: https://github.com/pkgcore/pkgcheck/issues/399

### CODE REVIEW

File: src/pkgcheck/checks/metadata.py [hunk 1]
```diff
@@ -804,14 +804,34 @@ class InvalidIdepend(results.MetadataError, results.VersionResult):
     attr = 'idepend'
 
 
+class MisplacedWeakBlocker(results.Warning, results.VersionResult):
+    """Weak blocker is in misplaced depends type.
```
Comment (thesamesam) [inline src/pkgcheck/checks/metadata.py:808]: Weak blocker is within a misplaced dependency eclass.

File: src/pkgcheck/checks/metadata.py [hunk 2]
```diff
@@ -804,14 +804,34 @@ class InvalidIdepend(results.MetadataError, results.VersionResult):
...
+
+    Weak blockers control whether we ignore file collisions at the point of
+    merge, so being defined at DEPEND or BDEPEND is wrong.
```
Comment (thesamesam) [inline src/pkgcheck/checks/metadata.py:811]:
```suggestion
    merge, so being (exclusively) defined in DEPEND or BDEPEND is wrong.
```

File: src/pkgcheck/checks/metadata.py [hunk 3]
```diff
@@ -804,14 +804,34 @@ class InvalidIdepend(results.MetadataError, results.VersionResult):
...
+
+    Note that in cases where the weak blocker is also defined at RDEPEND, this
```
Comment (thesamesam) [inline src/pkgcheck/checks/metadata.py:813]:
```suggestion
    Note that in cases where the weak blocker is also defined in RDEPEND, this
```

File: src/pkgcheck/checks/metadata.py [hunk 4]
```diff
@@ -804,14 +804,34 @@ class InvalidIdepend(results.MetadataError, results.VersionResult):
...
+    warning won't be triggered, as a leeway.
```
Comment (thesamesam) [inline src/pkgcheck/checks/metadata.py:814]:
```suggestion
    warning won't be triggered, to give leeway given this is a common ebuild pattern.
```

File: src/pkgcheck/checks/metadata.py [hunk 5]
```diff
@@ -889,6 +911,14 @@ def feed(self, pkg):
                         elif atom.slot_operator == '=':
                             yield BadDependency(
                                 attr, atom, '= slot operator used in blocker', pkg=pkg)
+                        elif not atom.blocks_strongly:
+                            weak_blocks[attr].add(atom)
+
+        for attr in ('depend', 'bdepend'):
```
Comment (thesamesam) [inline src/pkgcheck/checks/metadata.py:917]: Should we add IDEPEND in this block somewhere?

Comment (arthurzam) [reply src/pkgcheck/checks/metadata.py:918]: Will handle it tomorrow.
I was thinking that having a weak-blocker in `IDEPEND` was fine, but on second thought, why should it?

Comment (thesamesam) [reply src/pkgcheck/checks/metadata.py:918]: It's okay if it's in there along with other deps (but unlikely), but it doesn't make sense anyway, and it's like putting it in RDEPEND. We should treat it like the "In RDEPEND but also DEPEND" case.

Comment (arthurzam) [reply src/pkgcheck/checks/metadata.py:918]: Done, thanks for the clarification 😄

**Review by thesamesam — APPROVED** [review]

**Pull request merged**